File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.java
import java.util.*;

class Solution {
    static int N;
    static int[][] Dice;
    static boolean[] isSelected;
    static ArrayList<Integer> aScore = new ArrayList<>();
    static ArrayList<Integer> bScore = new ArrayList<>();
    static int MaxWin = 0;
    static boolean[] result;
    
    public int[] solution(int[][] dice) {
        N = dice.length;
        Dice = dice;
        result = new boolean[N];
        
        isSelected = new boolean[N];
        combi(0, 0);
        
        int[] answer = new int[N/2];
        int idx = 0;
        for(int i=0; i<N; i++){
            if(result[i]){
                answer[idx++] = i+1;
            }
        }
        
        return answer;
    }
    
    public void combi(int start, int cnt){
        if(cnt == N/2){
            ArrayList<Integer> aIdx = new ArrayList<>();
            ArrayList<Integer> bIdx = new ArrayList<>();
            
            for(int i=0; i<N; i++){
                if(isSelected[i]) aIdx.add(i);
                else bIdx.add(i);
            }
            
            aScore.clear();
            bScore.clear();
            
            // 현재의 주사위 조합에서 나올 수 있는 수의 합 모두 구하기
            dfs(0, aIdx, 0, true);
            dfs(0, bIdx, 0, false);
            
            Collections.sort(bScore);
            int win = 0;
            for (int aSum : aScore) {
                win += lowerBound(aSum);
                // bScore 중 하나의 aSum에 대해 점수가 낮은 것의 개수
            }
            
            if(MaxWin < win){
                MaxWin = win;
                result = isSelected.clone();
            }
            
            return;
        }
        
        for(int i=start; i<N; i++){
            isSelected[i] = true;
            combi(i+1, cnt+1);
            isSelected[i] = false;
        }
    }
    
    public void dfs(int cnt, ArrayList<Integer> idx, int sum, boolean a){
        // a = true라면 a의 경우의 수를 구하는 것
        
        if(cnt == N/2){
            if(a){
                aScore.add(sum);
                
            }else{
                bScore.add(sum);
                
            }
            
            return;
        }
        
        int diceIdx = idx.get(cnt);
        for(int i=0; i<6; i++){
            dfs(cnt+1, idx, sum + Dice[diceIdx][i], a);
        }
        
    }
    
    public int lowerBound(int x){
        int start = 0;
        int end = bScore.size()-1;
        int result = bScore.size();
        
        while(start <= end){
            int mid = (start + end) / 2;
            
            if(bScore.get(mid) >= x){
                result = mid;
                end = mid - 1;
                
            }else{
                start = mid + 1;
                
            }
            
        }
        
        return result;
    }
}

File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.py
def solution(dice):
    R = len(dice) // 2
    selected = [False]*len(dice)
    maxWin = 0
    ans = []
    
    def dfs(result, d, cnt, tot):
        if cnt == R:
            # 마지막 주사위까지 한 면을 골랐다면, 점수의 합을 결과 리스트에 넣는다.
            result.append(tot)
            return
        
        for i in range(6):
            # cnt번째 주사위의 6면 중 하나를 선택
            idx = d[cnt]
            dfs(result, d, cnt+1, tot+dice[idx][i])
            
    def lowerbound(a, bList):
        start = 0
        end = len(bList) - 1
        result = len(bList) # bList에서 처음으로 a 이상(>= a)이 되는 위치(index)
                            # 즉, a보다 작은 bList 값의 갯수를 알려준다.
        
        while start <= end:
            mid = (start+end) // 2
            
            if bList[mid] >= a:
                result = mid
                end = mid - 1
            else:
                start = mid + 1
                
        return result
    
    def combi(cnt, start):
        # A가 가져갈 주사위 선택
        if cnt == R:
            a = []
            b = []
            
            for i in range(len(selected)):
                if selected[i]:
                    a.append(i)
                else:
                    b.append(i)
                    
            # A가 주사위를 굴려 나올 수 있는 모든 경우의 수
            aList = []
            dfs(aList, a, 0, 0)
            
            # B가 주사위를 굴려 나올 수 있는 모든 경우의 수
            bList = []
            dfs(bList, b, 0, 0)
            
            # 각 A의 점수에 대해 A가 승리하는 경우의 수 세기
            bList.sort()
            
            win = 0
            for i in aList:
                win += lowerbound(i, bList)
                
            nonlocal maxWin, ans
            if maxWin < win:
                ans = []
                maxWin = win
                
                for i in range(len(selected)):
                    if selected[i]:
                        ans.append(i+1)
            
            return

        for i in range(start, len(dice)):
            selected[i] = True
            combi(cnt+1, i+1)
            selected[i] = False
            
    combi(0, 0)
    
    return ans